### ISSUE TITLE
Deal with fiscal period in the `CashCtrlLedger`

### ISSUE THREAD

**Issue by AlexTheWizardL** [macxred/pyledger#29]
### Description:
`CashCtrlLedger` package should automatically adapt to different fiscal periods when accessing and modifying ledger entries.
This requires knowledge sharing session with Lukas

**Comment by AlexTheWizardL** [comment]
### Possible Options for Handling Fiscal Periods in the CashCtrlLedger Package

### Option 1: Perfect Solution (Abstract Logic in PyLedger)
**Description**: Define abstract logic for managing fiscal periods in the `pyledger` package and implement a caching mechanism in the `cashctrl_api` package. The `CashCtrlLedger` package would implement the abstract methods to handle fiscal periods accordingly.  
**Benefits**:
- Follows a clean, modular approach by abstracting fiscal period handling.
- Provides a reusable, consistent mechanism for managing fiscal periods across different ledger implementations.
- Efficient caching in `cashctrl_api` will reduce redundant API calls, improving performance.

**Pitfalls**:
- This approach requires significant effort and time to design and implement.
- Adds complexity to the project by introducing new abstract methods and caching logic that isn't immediately necessary for the current problem.
- Fiscal periods aren’t an issue in `pyledger`, so introducing abstract logic at this level may be over-engineering for the current use case.

### Option 2: Define Fiscal Periods as Settings
**Description**: Define fiscal periods as settings in the form of a dictionary or configuration file. These settings would then be mirrored to CashCtrl via the API.  
**Benefits**:
- Quick and simple to implement.
- Can be integrated into the existing settings structure without requiring significant changes.
  
**Pitfalls**:
- This approach doesn’t align with the current DataFrame-based architecture.
- Mirroring fiscal periods through settings may introduce more complexity, as you would need to manage both settings and API logic separately.
- Less flexible and scalable, especially as new requirements for fiscal period handling arise.

### Option 3: Implement Logic for Current Needs (with Expandability)
**Description**: Define logic for creating fiscal periods when adding or updating ledger entries. When an entry’s date isn’t within an existing fiscal period, a new fiscal period for that year is created automatically. The logic will be implemented in a way that can be easily expanded in the future to handle more complex requirements, such as custom fiscal periods.  
**Benefits**:
- Simple and quick to implement for the current issue within `CashCtrlLedger`.
- Allows the system to function without requiring the user to manually create fiscal periods.
- Can be expanded in the future to handle more complex fiscal period requirements without a complete redesign.
- The design is aligned with the current DataFrame-based architecture and easy to replace later with accessor/mutator methods if needed.

**Pitfalls**:
- Users won’t be able to create custom fiscal periods through the terminal or API, relying on the GUI for this functionality.
- Although the logic is expandable, the initial implementation may not cover more advanced scenarios, such as fiscal periods spanning multiple years or custom periods.

---

### Chosen Solution: Option 3 (Implement Logic for Current Needs with Expandability)

**Why Chosen**:  
- The fiscal period issue is specific to `CashCtrlLedger`, and the overall project goals didn’t initially include handling fiscal periods as a core entity. Therefore, focusing on solving the problem at hand with future expandability in mind is the most pragmatic choice.
- Option 3 allows for a quick, effective solution that ensures ledger entries can be created and updated without fiscal period errors, while keeping the code flexible for future updates.
- This approach ensures that the code aligns with the current DataFrame-based architecture and doesn’t introduce unnecessary complexity at this stage.
- Although custom fiscal period creation through the terminal isn’t immediately possible, this feature can be added in the future without a complete overhaul of the design.

---

### Practical Solution Implementation

#### 1. **`cashctrl_api` Package - Getter for Fiscal Periods (with Caching)**

We’ll implement a caching mechanism to efficiently retrieve fiscal periods from CashCtrl, returning the result as a DataFrame. This reduces redundant API calls, especially when handling multiple ledger entries.

##### **Code for `cashctrl_api` Package**:

```python
def list_fiscal_periods(self) -> pd.DataFrame:
    """Lists fiscal periods from the CashCtrl API, with caching to reduce redundant API calls.

    Returns:
        pd.DataFrame: A DataFrame with fiscal periods.
    """
    if self._fiscal_period_cache is None or self._is_expired(self._fiscal_period_cache_time):
        # Fetch the fiscal periods from the API if cache is expired or doesn't exist
        fiscal_periods = pd.DataFrame(self.get("fiscalperiod/list.json")["data"])
        self._fiscal_period_cache = fiscal_periods
        self._fiscal_period_cache_time = datetime.now()
        
    return self._fiscal_period_cache
```

- **Caching**: Ensures that the fiscal periods are cached and invalidated after a set timeout to minimize redundant API calls.
- **Retrieval**: Calls the CashCtrl API only when the cache is expired or does not exist and returns a DataFrame.

---

#### 2. **`CashCtrlLedger` Package - Ensuring Fiscal Period Exists**

We’ll define a method in `CashCtrlLedger` that checks whether a fiscal period exists for a given ledger entry date. If the entry date falls outside the existing fiscal periods, a new fiscal period is created. This method will be called inside the `add_ledger_entry` and `update_ledger_entry` methods.

##### **Code for `CashCtrlLedger`**:

```python
class CashCtrlLedger:
    
    def __init__(self, client):
        self._client = client

    def ensure_fiscal_period_exists(self, entry_date: pd.Timestamp) -> None:
        """Ensure that a fiscal period exists for the given entry date.

        Args:
            entry_date (pd.Timestamp): The date of the ledger entry.
        """
        fiscal_periods = self._client.list_fiscal_periods()

        # Check if the entry date falls within any existing fiscal period
        fiscal_period_exists = any(
            pd.Timestamp(period["start"]) <= entry_date <= pd.Timestamp(period["end"])
            for _, period in fiscal_periods.iterrows()
        )

        if not fiscal_period_exists:
            # Create a new fiscal period for the year of the entry date
        fiscal_period_data = {
            "isCustom": True,
            "start": f"{entry_date.year}-01-01",
            "end": f"{entry_date.year}-12-31",
            "name": f"{entry_date.year}",  # Following CashCtrl naming convention
        }
        self._client.post("fiscalperiod/create.json", data=fiscal_period_data)
        self._client.invalidate_fiscal_period_cache()  # Invalidate cache after creation

    def add_ledger_entry(self, entry: pd.DataFrame) -> str:
        """Add a new ledger entry, ensuring the fiscal period exists.

        Args:
            entry (pd.DataFrame): The ledger entry DataFrame.

        Returns:
            str: The ID of the created ledger entry.
        """
        entry_date = pd.Timestamp(entry["date"].iloc[0])  # Assuming single-row DataFrame

        # Ensure that the fiscal period exists for the entry date
        self.ensure_fiscal_period_exists(entry_date)

        # Proceed with adding the ledger entry (existing logic)
        payload = self._map_ledger_entry(entry)
        res = self._client.post("journal/create.json", data=payload)
        self._client.invalidate_journal_cache()
        return str(res["insertId"])

    def update_ledger_entry(self, entry: pd.DataFrame) -> None:
        """Update a ledger entry, ensuring the fiscal period exists.

        Args:
            entry (pd.DataFrame): The ledger entry DataFrame.
        """
        entry_date = pd.Timestamp(entry["date"].iloc[0])  # Assuming single-row DataFrame

        # Ensure that the fiscal period exists for the updated entry date
        self.ensure_fiscal_period_exists(entry_date)

        # Proceed with updating the ledger entry (existing logic)
        payload = self._map_ledger_entry(entry)
        self._client.post("journal/update.json", data=payload)
```

- **`ensure_fiscal_period_exists`**: This method checks if the fiscal period exists for the entry date. If not, it creates a new fiscal period for that year.
- **Integration**: The method is called inside `add_ledger_entry` and `update_ledger_entry` to ensure the correct fiscal period is in place before adding or updating ledger entries.

---

### Summary:

1. **Chosen Solution (Option 3)**: Implement logic for current needs, with expandability for future custom fiscal period handling.
2. **`cashctrl_api` Package**: Implemented the `list_fiscal_periods` method with caching to reduce redundant API calls, returning a DataFrame.
3. **`CashCtrlLedger` Package**: Implemented the logic to ensure fiscal periods exist when adding or updating ledger entries, with room for future expansion to handle custom fiscal periods.
4. **Future Consideration**: The perfect solution (Option 1) can be implemented after the project completion or integrated within the project timeline if there is enough time. This would involve abstracting fiscal period logic to `pyledger` and implementing a more modular and reusable structure across ledger implementations.

**Comment by AlexTheWizardL** [comment]
Implemented in [PR #92](https://github.com/macxred/cashctrl_ledger/pull/92)
